File: sorting/mergesort.js
function merge_sort(arr, low, high) {
	if (low < high) {
		let mid = Math.floor((low + high)/2);
		merge_sort(arr, low, mid);
		merge_sort(arr, mid+1, high);
		merge(arr, low, mid, high);
	}
}
function merge(arr, low, mid, high) {
	//two temperory arrays for holding left and right sub arrays
	let ar1 = [], ar2 = [];
	//getting those subarrays with the help of slice method
	ar1 = arr.slice(low, mid + 1);
	ar2 = arr.slice(mid + 1, high + 1);
	let v1 = 0, v2 = 0, k = low;
	//putting them back into original array by comparing elements
	const ar1l = ar1.length, ar2l = ar2.length;
	while (v1 < ar1l && v2 < ar2l) {
		if (ar1[v1] < ar2[v2]) {
			arr[k++] = ar1[v1++];
		} else {
			arr[k++] = ar2[v2++];
		}
	}
	//adding remaining elements if subarrays are of unequal size
	while (v1 < ar1l) {
		arr[k++] = ar1[v1++];
	}
	while (v2 < ar2l) {
		arr[k++] = ar2[v2++];
	}
}
const arr = [];
const maxsize = 40;
for (let i = 0; i < maxsize; i++) {
	arr.push(Math.floor(Math.random() * 100));
}
console.log("before sorting");
console.log(arr);
merge_sort(arr, 0, maxsize - 1);
console.log("after sorting\n");
console.log(arr);

File: sorting/mergesort.c
#include "utility.h"

void merge(int *arr, int l, int m, int h) {
	//finding two temperory array sizes
	int ls = m - l + 1;
	int rs = h - m;
	//creating temp arrays
	int L[ls];
	int R[rs];
	//copying elements to subarrays
	for (int i = l, q = 0; i <= m; i++) {
		L[q++] = arr[i];
	}
	for (int i = m + 1, q = 0; i <= h; i++) {
		R[q++] = arr[i];
	}
	//merging arrays
	int v1 = 0, v2 = 0, k = l;
	//putting them back into original array by comparing elements
	while (v1 < ls && v2 < rs) {
		if (L[v1] <= R[v2]) {
			arr[k++] = L[v1++];
		} else {
			arr[k++] = R[v2++];
		}
	}
	//adding remaining elements if subarrays are of unequal size
	while (v1 < ls) {
		arr[k++] = L[v1++];
	}
	while (v2 < rs) {
		arr[k++] = R[v2++];
	}
}

void msort(int *arr, int l, int h) {
	if (l < h) {
		int mid = (l + h)/2;
		msort(arr, l, mid);
		msort(arr, mid+1, h);
		merge(arr, l, mid, h);
	}
}

int main() {
	srand(time(0));
	int upper = 100, lower = 0;
	int arr[MAX];
	for (int i = 0; i < MAX; i++) {
		arr[i] = (rand() % (upper - lower + 1)) + lower;
	}

	printarray(arr, MAX);
	msort(arr, 0, MAX - 1);
	printarray(arr, MAX);
}

